pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+		return nil, err
+	}
+	return sc, nil
+}
+
+func adConfiguration(sc *config.ScaledContext) (*v3.ActiveDirectoryConfig, error) {
+	authConfigs := sc.Management.AuthConfigs("")
+	secrets := sc.Core.Secrets("")
+
+	authConfigObj, err := authConfigs.ObjectClient().UnstructuredClient().Get("activedirectory", metav1.GetOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] failed to obtain activedirecotry authConfigObj: %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+
+	u, ok := authConfigObj.(runtime.Unstructured)
+	if !ok {
+		logrus.Errorf("[%v] failed to retrieve ActiveDirectoryConfig, cannot read k8s Unstructured data %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+	storedADConfigMap := u.UnstructuredContent()
+
+	storedADConfig := &v3.ActiveDirectoryConfig{}
+	err = mapstructure.Decode(storedADConfigMap, storedADConfig)
+	if err != nil {
+		logrus.Debugf("[%v] errors while decoding stored AD config: %v", migrateAdUserOperation, err)
+	}
+
+	metadataMap, ok := storedADConfigMap["metadata"].(map[string]interface{})
+	if !ok {
+		logrus.Errorf("[%v] failed to retrieve ActiveDirectoryConfig, (second step), cannot read k8s Unstructured data %v", migrateAdUserOperation, err)
+	}
+
+	typemeta := &metav1.ObjectMeta{}
+	err = mapstructure.Decode(metadataMap, typemeta)
+	if err != nil {
+		logrus.Debugf("[%v] errors while decoding typemeta: %v", migrateAdUserOperation, err)
+	}
+
+	storedADConfig.ObjectMeta = *typemeta
+
+	logrus.Debugf("[%v] Should in theory have ActiveDirectory config data? Let's check!", migrateAdUserOperation)
+	logrus.Debugf("[%v] AD Service Account User: %v", migrateAdUserOperation, storedADConfig.ServiceAccountUsername)
+
+	if storedADConfig.ServiceAccountPassword != "" {
+		value, err := common.ReadFromSecret(secrets, storedADConfig.ServiceAccountPassword,
+			strings.ToLower(v3client.ActiveDirectoryConfigFieldServiceAccountPassword))
+		if err != nil {
+			return nil, err
+		}
+		storedADConfig.ServiceAccountPassword = value
+	}
+
+	return storedADConfig, nil
+}
+
+func newCAPool(cert string) (*x509.CertPool, error) {
+	pool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+	pool.AppendCertsFromPEM([]byte(cert))
+	return pool, nil
+}
+
+func ldapConnection(config *v3.ActiveDirectoryConfig) (*ldapv3.Conn, error) {
+	caPool, err := newCAPool(config.Certificate)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create caPool: %v", err)
+	}
+
+	servers := config.Servers
+	TLS := config.TLS
+	port := config.Port
+	connectionTimeout := config.ConnectionTimeout
+	startTLS := config.StartTLS
+
+	ldapConn, err := ldap.NewLDAPConn(servers, TLS, startTLS, port, connectionTimeout, caPool)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccountUsername := ldap.GetUserExternalID(config.ServiceAccountUsername, config.DefaultLoginDomain)
+	err = ldapConn.Bind(serviceAccountUsername, config.ServiceAccountPassword)
```

Comment:
... that's a good question. I guess logically the service account is _supposed_ to be able to see all the users, but it's possible that there's a scenario where the user can see themselves, but for whatever reason the SA can't see them. Hrm... in that case, perhaps we should attempt to bind as the user we are looking up before we declare them to be truly missing.

Reply:
:man_facepalming: Which we can't do because we don't _have_ the local user's password in this context, and it's inappropriate to expect that the admin running the script will know it. So ... our alternatives are nonexistent.

This has ramifications for _documentation_, since we need to call out that the "missing users" functionality of the script is limited by the visibility of the service account. Helpfully this isn't a gigantic problem as "delete those missing users" is non-default behavior, and anyone using it is **strongly** advised to first do a dry-run and make sure the results look sensible.